### PR TITLE
Fixed issue #338 bigtable::DeleteFromColumn* should take std::duration<> parameters.

### PR DESCRIPTION
**Pull request by roopak-qlogic**

### CODE REVIEW

File: bigtable/client/mutations.cc [hunk 1]
```diff
@@ -40,14 +40,14 @@ Mutation SetCell(std::string family, std::string column, std::string value) {
 }
 
 Mutation DeleteFromColumn(std::string family, std::string column,
-                          std::int64_t timestamp_begin,
-                          std::int64_t timestamp_end) {
+                          std::chrono::duration<std::int64_t> timestamp_begin,
```
Comment (coryan) [inline bigtable/client/mutations.cc]: Please look at how we use `std::chrono::duration<>` in `bigtable::Filters`.

File: bigtable/tests/mutations_integration_test.cc [hunk 2]
```diff
@@ -378,7 +388,7 @@ TEST_F(MutationIntegrationTest, DeleteFromColumnStartingFromTest) {
 
   // Create records
   CreateCells(*table, created_cells);
-  std::int64_t time_begin = 1000;
+  std::chrono::duration<std::int64_t> time_begin(1000);
```
Comment (coryan) [inline bigtable/tests/mutations_integration_test.cc]: Ideally we would like to say something like:

```C++
using namespace std::chrono_literals; // assumes C++14
auto time_begin = 1ms;
```

And the library automatically knows to what units this needs to be converted inside.

**Comment by roopak-qlogic** [comment]
Working on, one of the test case, which failed against Bigtable emulator

**Comment by roopak-qlogic** [comment]
build issues resolved

File: bigtable/client/mutations.h [hunk 3]
```diff
@@ -74,17 +75,54 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
  */
 
 /// Delete only within the timestamp range provided.
+template <typename Rep1, typename Period1, typename Rep2, typename Period2>
```
Comment (coryan) [inline bigtable/client/mutations.h]: Please include Doxygen documentation for `Rep1`, `Period1` and company.  You can use the examples in the `bigtable/client/filters.h` file.

File: bigtable/client/mutations.h [hunk 4]
```diff
@@ -74,17 +75,54 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+      duration_cast<microseconds>(timestamp_begin).count());
+  d.mutable_time_range()->set_end_timestamp_micros(
+      duration_cast<microseconds>(timestamp_end).count());
+  return m;
+}
+
 /// Delete starting from, and including, @a timestamp_begin.
-Mutation DeleteFromColumnStartingFrom(std::string family, std::string column,
-                                      std::int64_t timestamp_begin);
+template <typename Rep1, typename Period1>
```
Comment (coryan) [inline bigtable/client/mutations.h]: Same here, please add Doxygen documentation.  Also use `Rep` vs. `Rep1` (and `Period` vs. `Period`) in this case.

File: bigtable/client/mutations.h [hunk 5]
```diff
@@ -74,17 +75,54 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+  d.set_column_qualifier(std::move(column));
+  d.mutable_time_range()->set_start_timestamp_micros(
+      duration_cast<microseconds>(timestamp_begin).count());
+  return m;
+}
+
 /// Delete up to, but excluding, @a timestamp_end.
-Mutation DeleteFromColumnEndingAt(std::string family, std::string column,
-                                  std::int64_t timestamp_end);
+template <typename Rep2, typename Period2>
```
Comment (coryan) [inline bigtable/client/mutations.h]: Ditto.

File: bigtable/client/idempotent_mutation_policy_test.cc [hunk 6]
```diff
@@ -21,7 +21,7 @@ TEST(IdempotentMutationPolicyTest, Simple) {
   using namespace bigtable::chrono_literals;
   auto policy = bigtable::DefaultIdempotentMutationPolicy();
   EXPECT_TRUE(policy->is_idempotent(
-      bigtable::DeleteFromColumn("fam", "col", 0, 10).op));
+      bigtable::DeleteFromColumn("fam", "col", 0_ms, 10_ms).op));
```
Comment (coryan) [inline bigtable/client/idempotent_mutation_policy_test.cc]: Why did you change the values from microseconds to milliseconds in these tests?  Just curious, it does not matter but I wonder if something interesting was going one.

File: bigtable/client/mutations.h [hunk 7]
```diff
@@ -22,6 +22,7 @@
 #include <google/bigtable/v2/data.pb.h>
 #include <grpc++/grpc++.h>
```
Comment (coryan) [inline bigtable/client/mutations.h]: Remove this blank line please.

File: bigtable/client/mutations.h [hunk 8]
```diff
@@ -22,6 +22,7 @@
...
 
+#include <chrono>
 #include <type_traits>
```
Comment (coryan) [inline bigtable/client/mutations.h]: Please remove this blank line.

File: bigtable/tests/mutations_integration_test.cc [hunk 9]
```diff
@@ -255,17 +256,18 @@ TEST_F(MutationIntegrationTest, DeleteFromColumnForReversedTimestampRangeTest) {
...
+
 #if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
   // Try to delete the columns with an invalid range:
-  EXPECT_THROW(
-      table->Apply(bigtable::SingleRowMutation(
-          key, bigtable::DeleteFromColumn(column_family2, "c2", 4000, 2000))),
-      bigtable::PermanentMutationFailure);
+  EXPECT_THROW(table->Apply(bigtable::SingleRowMutation(
+                   key, bigtable::DeleteFromColumn(column_family2, "c2",
+                                                   4000_ms, 2000_ms))),
```
Comment (coryan) [inline bigtable/tests/mutations_integration_test.cc]: Why not use `4_ms` and `2_ms` here?

File: bigtable/client/idempotent_mutation_policy_test.cc (hunk 6 above)
Comment (roopak-qlogic) [reply bigtable/client/idempotent_mutation_policy_test.cc]: I thought number we passing, is default milliseconds, so I added _ms.
So if default we pass as micro seconds, then we can use  10_us instead of 10_ms, am I correct ?

Comment (coryan) [reply bigtable/client/idempotent_mutation_policy_test.cc]: Yup, the previous values (as raw integers) represented microseconds. Using `10_us` would require less changes in the tests.

Comment (roopak-qlogic) [reply bigtable/client/idempotent_mutation_policy_test.cc]: OK, then I will change every where _us,  where I have used _ms for raw integers

Comment (coryan) [reply bigtable/client/idempotent_mutation_policy_test.cc]: You can use the units that are most natural for each test.  For example, if the test used to say `2000`, that can now be represented as `2_ms` or `2000_us`.  The former is probably more readable.  That is the whole point of supporting `std::duration<>`: to make the code easy to understand.

File: bigtable/tests/mutations_integration_test.cc (hunk 9 above)
Comment (roopak-qlogic) [reply bigtable/tests/mutations_integration_test.cc]: Yes like your earlier comment, due to my  wrong understanding about default raw integer is millisecond, I did mistake here as well.
so here 4000_us and 2000_us is more obvious, and also I have to remove all the values multiplied by 1000, to make code less change and more readable.
Thanks for comments and suggestion.

File: bigtable/client/mutations.h [hunk 10]
```diff
@@ -63,28 +63,136 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
  * cells in the given column family and column within the given
- * timestamp range.
+ * timestamp in the range.
+ *
+ * The function accepts any instantiation of `std::chrono::duration<>` for the
+ * @p timestamp_begin and @p timestamp_end parameters.  For example:
+ *
+ * @code
+ * using namespace std::chrono_literals; // C++14
+ * bigtable::DeleteFromColumn("fam", "col", 0_us, 10_us)
```
Comment (coryan) [inline bigtable/client/mutations.h]: In C++14 the literals do not require a leading `_`.

File: bigtable/client/mutations.h [hunk 11]
```diff
@@ -63,28 +63,136 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
  * The ending timestamp is exclusive, while the beginning timestamp is
  * inclusive.  That is, the interval is [@p timestamp_begin, @p timestamp_end).
  * The value 0 is special and treated as "unbounded" for both the begin and end
  * endpoints of the time range.  The Cloud Bigtable server rejects invalid and
  * empty ranges, i.e., any range where the endpoint is smaller or equal than to
  * the initial endpoint unless either endpoint is 0.
+ *
+ * @tparam Rep1 a placeholder to match the Rep tparam for @p timestamp_begin
+ * type,
+ *     the semantics of this template parameter are documented in
```
Comment (coryan) [inline bigtable/client/mutations.h]: please indent `type,` and reflow the paragraph.

File: bigtable/client/mutations.h [hunk 12]
```diff
@@ -63,28 +63,136 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+ *     `std::chrono::duration<>` (in brief, the underlying arithmetic type
+ *     used to store the number of ticks), for our purposes it is simply a
+ *     formal parameter.
+ * @tparam Rep2 similar formal parameter for the type of @p timestamp_end.
+ * @tparam Period1 a placeholder to match the Period tparam for @p
+ * timestamp_begin
+ *     type, the semantics of this template parameter are documented in
```
Comment (coryan) [inline bigtable/client/mutations.h]: Same here.

File: bigtable/client/mutations.h [hunk 13]
```diff
@@ -63,28 +63,136 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+ * The function accepts any instantiation of `std::chrono::duration<>` for the
+ * @p timestamp_begin For example:
+ *
+ * @code
+ * using namespace std::chrono_literals; // C++14
+ * bigtable::DeleteFromColumn("fam", "col", 10_us)
+ * @endcode
+ *
+ * @tparam Rep1 a placeholder to match the Rep tparam for @p timestamp_begin
+ * type,
```
Comment (coryan) [inline bigtable/client/mutations.h]: and here.

File: bigtable/client/mutations.h [hunk 14]
```diff
@@ -63,28 +63,136 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+ *     the semantics of this template parameter are documented in
+ *     `std::chrono::duration<>` (in brief, the underlying arithmetic type
+ *     used to store the number of ticks), for our purposes it is simply a
+ *     formal parameter.
+ * @tparam Period1 a placeholder to match the Period tparam for @p
+ * timestamp_begin
```
Comment (coryan) [inline bigtable/client/mutations.h]: and here.

File: bigtable/client/mutations.h [hunk 15]
```diff
@@ -63,28 +63,136 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+ * @name The following functions create a mutation that deletes all the
+ * cells in the given column family and column, Delete up to @a timestamp_end,
+ * but excluding, @a timestamp_end.
+ *
+ * The function accepts any instantiation of `std::chrono::duration<>` for the
+ * @p timestamp_end For example:
+ *
+ * @code
+ * using namespace std::chrono_literals; // C++14
+ * bigtable::DeleteFromColumn("fam", "col", 10_us)
```
Comment (coryan) [inline bigtable/client/mutations.h]: No leading `_` if you are using C++14.

File: bigtable/client/mutations.h [hunk 16]
```diff
@@ -62,37 +62,39 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+ * empty ranges, i.e., any range where the end point is smaller or equal than to
+ * the initial end point unless either end point is 0.
  *
  * @tparam Rep1 a placeholder to match the Rep tparam for @p timestamp_begin
- * type,
- *     the semantics of this template parameter are documented in
- *     `std::chrono::duration<>` (in brief, the underlying arithmetic type
- *     used to store the number of ticks), for our purposes it is simply a
- *     formal parameter.
+ * type.The semantics of this template parameter are documented in
```
Comment (coryan) [inline bigtable/client/mutations.h]: s/type.The/type. /

File: bigtable/client/mutations.h [hunk 17]
```diff
@@ -122,21 +124,21 @@ Mutation DeleteFromColumn(std::string family, std::string column,
...
+ * bigtable::DeleteFromColumn("fam", "col", 10us)
  * @endcode
  *
  * @tparam Rep1 a placeholder to match the Rep tparam for @p timestamp_begin
- * type,
- *     the semantics of this template parameter are documented in
- *     `std::chrono::duration<>` (in brief, the underlying arithmetic type
- *     used to store the number of ticks), for our purposes it is simply a
- *     formal parameter.
+ * type.The semantics of this template parameter are documented in
```
Comment (coryan) [inline bigtable/client/mutations.h]: Same.

File: bigtable/client/mutations.h [hunk 18]
```diff
@@ -122,21 +124,21 @@ Mutation DeleteFromColumn(std::string family, std::string column,
...
+ * used to store the number of ticks), for our purposes it is simply a
+ * formal parameter.
+ *
  * @tparam Period1 a placeholder to match the Period tparam for @p
  * timestamp_begin
- *     type, the semantics of this template parameter are documented in
- *     `std::chrono::duration<>` (in brief, the length of the tick in seconds,
- *     expressed as a `std::ratio<>`), for our purposes it is simply a formal
- *     parameter.
+ * type.the semantics of this template parameter are documented in
```
Comment (coryan) [inline bigtable/client/mutations.h]: Here `s/type.the/type. The/`

File: bigtable/client/mutations.h [hunk 19]
```diff
@@ -163,19 +165,20 @@ Mutation DeleteFromColumnStartingFrom(
...
+ * `std::chrono::duration<>` (in brief, the underlying arithmetic type
+ * used to store the number of ticks), for our purposes it is simply a
+ * formal parameter.
+ *
  * @tparam Period2 a placeholder to match the Period tparam for @p timestamp_end
- *     type, the semantics of this template parameter are documented in
- *     `std::chrono::duration<>` (in brief, the length of the tick in seconds,
- *     expressed as a `std::ratio<>`), for our purposes it is simply a formal
- *     parameter.
+ * type.The semantics of this template parameter are documented in
```
Comment (coryan) [inline bigtable/client/mutations.h]: Also missing space after the period here.

File: bigtable/client/mutations.h [hunk 20]
```diff
@@ -62,29 +62,140 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+ * @p timestamp_begin and @p timestamp_end parameters.  For example:
+ *
+ * @code
+ * using namespace std::chrono_literals; // C++14
+ * bigtable::DeleteFromColumn("fam", "col", 0us, 10us)
+ * @endcode
  *
  * The ending timestamp is exclusive, while the beginning timestamp is
- * inclusive.  That is, the interval is [@p timestamp_begin, @p timestamp_end).
+ * inclusive.  That is, the interval is [@p timestamp_begin, @p timestamp_end].
```
Comment (coryan) [inline bigtable/client/mutations.h]: Is the timestamp range inclusive of the endpoint?  If so, then the typical interval notation is `[a, b)` as opposed to `[a, b]`, right?

https://www.basic-mathematics.com/interval-notation.html

File: bigtable/client/mutations.h [hunk 21]
```diff
@@ -62,29 +62,140 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
  * The value 0 is special and treated as "unbounded" for both the begin and end
- * endpoints of the time range.  The Cloud Bigtable server rejects invalid and
- * empty ranges, i.e., any range where the endpoint is smaller or equal than to
- * the initial endpoint unless either endpoint is 0.
+ * points of the time range.  The Cloud Bigtable server rejects invalid and
+ * empty ranges, i.e., any range where the end point is smaller or equal than to
```
Comment (coryan) [inline bigtable/client/mutations.h]: We have been using "endpoint:, one word, keep it like that please.

File: bigtable/client/mutations.h [hunk 22]
```diff
@@ -62,29 +62,140 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+ * @tparam Rep1 a placeholder to match the Rep tparam for @p timestamp_begin
+ * type.The semantics of this template parameter are documented in
+ * std::chrono::duration<>` (in brief, the underlying arithmetic type
+ * used to store the number of ticks), for our purposes it is simply a
+ * formal parameter.
+ *
+ * @tparam Rep2 similar formal parameter for the type of @p timestamp_end.
+ *
+ * @tparam Period1 a placeholder to match the Period tparam for @p
+ * timestamp_begin
```
Comment (coryan) [inline bigtable/client/mutations.h]: still needs reflowing here.

File: bigtable/client/mutations.h [hunk 23]
```diff
@@ -62,29 +62,140 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+ * @endcode
+ *
+ * @tparam Rep1 a placeholder to match the Rep tparam for @p timestamp_begin
+ * type.The semantics of this template parameter are documented in
+ * `std::chrono::duration<>` (in brief, the underlying arithmetic type
+ * used to store the number of ticks), for our purposes it is simply a
+ * formal parameter.
+ *
+ * @tparam Period1 a placeholder to match the Period tparam for @p
+ * timestamp_begin
```
Comment (coryan) [inline bigtable/client/mutations.h]: and here.  Also missing space after the `.`.

Comment (roopak-qlogic) [reply bigtable/client/mutations.h]: After indention of  doxygen, I use clang-format-4.0 for formatting, before commit.
so some how, clang-format-4.0 changed the indentation again.
will check it

File: bigtable/client/mutations.h [hunk 24]
```diff
@@ -62,29 +62,138 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+ * @tparam Rep1 a placeholder to match the Rep tparam for @p timestamp_begin
+ * type. The semantics of this template parameter are documented in
+ * std::chrono::duration<>` (in brief, the underlying arithmetic type
+ * used to store the number of ticks), for our purposes it is simply a
+ * formal parameter.
+ *
+ * @tparam Rep2 similar formal parameter for the type of @p timestamp_end.
+ *
+ * @tparam Period1 a placeholder to match the Period tparam for
+ * @p timestamp_begin type. the semantics of this template parameter are
```
Comment (coryan) [inline bigtable/client/mutations.h]: s/type. the/type. The/

File: bigtable/client/mutations.h [hunk 25]
```diff
@@ -62,29 +62,138 @@ constexpr std::int64_t ServerSetTimestamp() { return -1; }
...
+ * @endcode
+ *
+ * @tparam Rep1 a placeholder to match the Rep tparam for @p timestamp_begin
+ * type. The semantics of this template parameter are documented in
+ * `std::chrono::duration<>` (in brief, the underlying arithmetic type
+ * used to store the number of ticks), for our purposes it is simply a
+ * formal parameter.
+ *
+ * @tparam Period1 a placeholder to match the Period tparam for @p
+ * timestamp_begin type. the semantics of this template parameter
```
Comment (coryan) [inline bigtable/client/mutations.h]: s/type. the/type. Type/

File: bigtable/client/mutations.h (hunk 24 above)
Comment (roopak-qlogic) [reply bigtable/client/mutations.h]: If I am correct then,  "type. the semantics" should be "type. The semantics" right ?

Comment (coryan) [reply bigtable/client/mutations.h]: Yes.
